File: calc-refactor/typescript/src/calculator.ts
// Identical byte-for-byte across C#, TypeScript, and Python.
// See ../../SCENARIOS.md for the characterization contract.

export const Keys = {
  equals: '=',
  clear: 'C',
} as const;

export const DisplayStrings = {
  zero: '0',
  error: 'Error',
} as const;

export const ErrorMessages = {
  unknownKey: (key: string): string => `unknown key: ${key}`,
} as const;

type Operator = '+' | '-' | '*' | '/';

function isDigit(c: string): boolean {
  return c.length === 1 && c >= '0' && c <= '9';
}

function isOperator(c: string): c is Operator {
  return c === '+' || c === '-' || c === '*' || c === '/';
}

// Integer division truncating toward zero — matches C# `/` and Python's
// `int(a/b)` rounding, so "7/2=" shows "3" and "-7/2=" shows "-3" in all
// three languages.
function truncDiv(a: number, b: number): number {
  return Math.trunc(a / b);
}

export class Calculator {
  private _display: string = DisplayStrings.zero;
  private left = 0;
  private pendingOperator: Operator | null = null;
  private rememberedRight = 0;
  private hasRememberedRight = false;
  private enteringOperand = false;
  private justEvaluated = false;
  private inError = false;

  get display(): string {
    return this._display;
  }

  press(key: string): void {
    if (this.inError) {
      if (key === Keys.clear) this.reset();
      return;
    }

    if (isDigit(key)) { this.pressDigit(key); return; }
    if (isOperator(key)) { this.pressOperator(key); return; }
    if (key === Keys.equals) { this.pressEquals(); return; }
    if (key === Keys.clear) { this.reset(); return; }

    throw new Error(ErrorMessages.unknownKey(key));
  }

  private pressDigit(key: string): void {
    if (!this.enteringOperand || this.justEvaluated) {
      if (this.justEvaluated) {
        // A digit after `=` starts a fresh calculation — drop the
        // accumulator and the pending operator so the new operand
        // is not folded into the remembered state.
        this.left = 0;
        this.pendingOperator = null;
        this.hasRememberedRight = false;
      }
      this._display = key;
      this.enteringOperand = true;
      this.justEvaluated = false;
      return;
    }

    this._display = this._display === DisplayStrings.zero ? key : this._display + key;
  }

  private pressOperator(key: Operator): void {
    if (this.enteringOperand || this.justEvaluated) {
      if (this.pendingOperator !== null && this.enteringOperand) {
        this.left = this.apply(this.left, this.pendingOperator, parseInt(this._display, 10));
        this._display = this.left.toString();
        if (this.inError) return;
      } else {
        this.left = parseInt(this._display, 10);
      }
      this.pendingOperator = key;
      this.enteringOperand = false;
      this.justEvaluated = false;
      this.hasRememberedRight = false;
      return;
    }

    // Two operators in a row — the legacy `1++2` path.
    this.enterError();
  }

  private pressEquals(): void {
    if (this.pendingOperator === null) return;

    const right = this.hasRememberedRight && !this.enteringOperand
      ? this.rememberedRight
      : parseInt(this._display, 10);

    const result = this.apply(this.left, this.pendingOperator, right);
    if (this.inError) return;

    this.rememberedRight = right;
    this.hasRememberedRight = true;
    this.left = result;
    this._display = result.toString();
    this.enteringOperand = false;
    this.justEvaluated = true;
  }

  private apply(left: number, op: Operator, right: number): number {
    switch (op) {
      case '+': return left + right;
      case '-': return left - right;
      case '*': return left * right;
      case '/':
        if (right === 0) { this.enterError(); return 0; }
        return truncDiv(left, right);
    }
  }

  private enterError(): void {
    this.inError = true;
    this._display = DisplayStrings.error;
  }

  private reset(): void {
    this._display = DisplayStrings.zero;
    this.left = 0;
    this.pendingOperator = null;
    this.rememberedRight = 0;
    this.hasRememberedRight = false;
    this.enteringOperand = false;
    this.justEvaluated = false;
    this.inError = false;
  }
}

File: calc-refactor/python/src/calc_refactor/calculator.py
from typing import Optional


# Identical byte-for-byte across C#, TypeScript, and Python.
# See ../../SCENARIOS.md for the characterization contract.
class KEYS:
    EQUALS = "="
    CLEAR = "C"


class DISPLAY:
    ZERO = "0"
    ERROR = "Error"


class ERRORS:
    @staticmethod
    def unknown_key(key: str) -> str:
        return f"unknown key: {key}"


_OPERATORS = frozenset({"+", "-", "*", "/"})


def _is_digit(c: str) -> bool:
    return len(c) == 1 and "0" <= c <= "9"


def _trunc_div(a: int, b: int) -> int:
    # Truncate toward zero — matches C# `/` and TS Math.trunc(a/b).
    # Python's `//` floors toward -inf, so we correct the sign here.
    return int(a / b) if b != 0 else 0


class Calculator:
    def __init__(self) -> None:
        self._display: str = DISPLAY.ZERO
        self._left: int = 0
        self._pending_operator: Optional[str] = None
        self._remembered_right: int = 0
        self._has_remembered_right: bool = False
        self._entering_operand: bool = False
        self._just_evaluated: bool = False
        self._in_error: bool = False

    @property
    def display(self) -> str:
        return self._display

    def press(self, key: str) -> None:
        if self._in_error:
            if key == KEYS.CLEAR:
                self._reset()
            return

        if _is_digit(key):
            self._press_digit(key)
            return
        if key in _OPERATORS:
            self._press_operator(key)
            return
        if key == KEYS.EQUALS:
            self._press_equals()
            return
        if key == KEYS.CLEAR:
            self._reset()
            return

        raise ValueError(ERRORS.unknown_key(key))

    def _press_digit(self, key: str) -> None:
        if not self._entering_operand or self._just_evaluated:
            if self._just_evaluated:
                # A digit after `=` starts a fresh calculation — drop the
                # accumulator and the pending operator so the new operand
                # is not folded into the remembered state.
                self._left = 0
                self._pending_operator = None
                self._has_remembered_right = False
            self._display = key
            self._entering_operand = True
            self._just_evaluated = False
            return

        self._display = key if self._display == DISPLAY.ZERO else self._display + key

    def _press_operator(self, key: str) -> None:
        if self._entering_operand or self._just_evaluated:
            if self._pending_operator is not None and self._entering_operand:
                self._left = self._apply(self._left, self._pending_operator, int(self._display))
                self._display = str(self._left)
                if self._in_error:
                    return
            else:
                self._left = int(self._display)
            self._pending_operator = key
            self._entering_operand = False
            self._just_evaluated = False
            self._has_remembered_right = False
            return

        # Two operators in a row — the legacy `1++2` path.
        self._enter_error()

    def _press_equals(self) -> None:
        if self._pending_operator is None:
            return

        right = (
            self._remembered_right
            if self._has_remembered_right and not self._entering_operand
            else int(self._display)
        )

        result = self._apply(self._left, self._pending_operator, right)
        if self._in_error:
            return

        self._remembered_right = right
        self._has_remembered_right = True
        self._left = result
        self._display = str(result)
        self._entering_operand = False
        self._just_evaluated = True

    def _apply(self, left: int, op: str, right: int) -> int:
        if op == "+":
            return left + right
        if op == "-":
            return left - right
        if op == "*":
            return left * right
        if op == "/":
            if right == 0:
                self._enter_error()
                return 0
            return _trunc_div(left, right)
        raise AssertionError(f"unreachable operator: {op}")

    def _enter_error(self) -> None:
        self._in_error = True
        self._display = DISPLAY.ERROR

    def _reset(self) -> None:
        self._display = DISPLAY.ZERO
        self._left = 0
        self._pending_operator = None
        self._remembered_right = 0
        self._has_remembered_right = False
        self._entering_operand = False
        self._just_evaluated = False
        self._in_error = False
